### PR TITLE
Add "Overview"

### PR DESCRIPTION
**Pull request by kaycebasques**
Just a braindump of how I would conceptually explain WB's value to a potential user

overview

### CODE REVIEW

**Comment by kaycebasques** [comment]
I also changed the tagline to `JavaScript Libraries For Offline Web Apps` for dramatic effect :)

**Comment by jeffposnick** [comment]
Thanks for working on this!

I haven't digested the full set of text yet, but re: the "JavaScript Libraries For Offline Web Apps" tagline—I don't want to give the impression that these libraries are only for "offline" use cases. Going cache-first even when a device is online leads to reliable, fast performance. "Offline-First" would be my preference over "Offline", or just keep it "Progressive".

**Comment by kaycebasques** [comment]
My big issue with "JS Libraries for PWAs" is that you look at the definition for PWA, and it has sections on being responsive, add to homescreen, serving over HTTPS, being indexed by Google, etc. Workbox has nothing to do with those, so it's way too broad to define it as a PWA library.

The bit about it making it sound like it's only for offline is a good point though. You may never expect users to go fully offline, maybe you're just caching app shell stuff for a faster UX

I do think that we should continue to really think about a tagline that captures the value of Workbox in a sentence. We can create something that creates much more meaning than "PWA"

**Comment by gauntface** [comment]
"JS Libraries that make service workers easy"? Ultimately I'd like workbox to contain a library or two for push in the near future.

File: Gemfile.lock
Comment (gauntface) [inline Gemfile.lock:1]: I don't think this file should be commited.

File: src/index.md
Comment (gauntface) [inline src/index.md]: Maybe link up background workers to: https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API

File: src/index.md
Comment (gauntface) [inline src/index.md]: Personal preference:


> The logic for determining whether to fetch resources from the cache or
> from the network is called a **caching strategy**

File: src/index.md
Comment (gauntface) [inline src/index.md]: "...and which one you use depends on the types or resources, how they are used and the importance of the resource."

File: src/index.md
Comment (gauntface) [inline src/index.md]: `workbox-build` and `workbox-cli` **can** generate a complete service worker for

File: src/index.md
Comment (gauntface) [inline src/index.md]: Not sure how to weave it in, but one of the things workbox does here is that it updates the resources efficiently (i.e. it only updates resources that have changed, and doesn't re-download all resources when there is a new service worker. It's a subtle feature of workbox and may not be worth mentioning.

File: src/index.md
Comment (gauntface) [inline src/index.md]: I might add a section header here, it wasn't entirely sure that you've moved from generating a full service worker with workbox-build or workbox-cli to injecting the precaching into a custom sw.

**Comment by gauntface** [comment]
@kaycebasques thank you so much for this - it's awesome!!

I've added comments but they are all personal. Where do you think this should live on the microsite? I'm not sure it's a good idea to have a wall of text as the first thing developers see, but again that's personal preference and if you think it's a good idea then let's do it.

**Comment by kaycebasques** [comment]
> Where do you think this should live on the microsite? 

This was just a hack to get it displayed on the site, I'm not suggesting to add it to the homepage like this

Although, since users have said that they look at the homepage and have no idea what Workbox is about, it may be worthwhile to add some short (1-3 paragraph) overview on the homepage

**Comment by gauntface** [comment]
+1 to adding like a TL;DR of this for home page and linking up to somewhere for this. Let me know if you want any help figuring out a home for this.

File: Gemfile.lock
Comment (jeffposnick) [reply Gemfile.lock:1]: My understanding is that it should be committed, for the same reason that we recently started committing our `package-lock.json`: https://stackoverflow.com/a/4151540/385997

File: src/index.md
Comment (jeffposnick) [inline src/index.md]: This would benefit from us getting on board with a consistent CDN URL (`unpkg` or otherwise) and then promoting that.

**Comment by addyosmani** [comment]
> Although, since users have said that they look at the homepage and have no idea what Workbox is about, it may be worthwhile to add some short (1-3 paragraph) overview on the homepage

This is awesome. I think it would be immensely valuable for us to add that shorter summary overview on the homepage and as Matt suggested linking up to a longer overview doc. Would you be up for taking a stab at that @kaycebasques? 🍪

File: src/index.md
Comment (addyosmani) [reply src/index.md]: 👍 for contextually linking out to MDN docs for some of these concepts newcomers might wish to dig into further.

**Comment by kaycebasques** [comment]
Hey @jeffposnick @gauntface @addyosmani I iterated on the Overview and tried to get it all to a shippable state

**Comment by gauntface** [comment]
For some reason the overview page doesn't build on prod builds, will dig into why soon

**Comment by kaycebasques** [comment]
@gauntface it's because I'm using a pretty URL in the nav... we can change it to `overview.html` or see if Jekyll has a flag that we can switch

**Comment by kaycebasques** [comment]
I'd prefer option 2 (enabling pretty URLs)

File: src/index.md
Comment (addyosmani) [inline src/index.md]: Looks good. I've seen `store` and `storage` used more often when describing the action of offline-caching data. I wonder if that would make more sense here than `save`? Your call.

File: src/index.md
Comment (addyosmani) [inline src/index.md]: Do you think we might want to be clear that we're talking about repeat-view performance or future visit performance rather than first view load perf in general?

File: src/overview.md
Comment (addyosmani) [inline src/overview.md]: Would you be open to changing this to `locally is called the (Service Worker) **cache**`? I think the simplicity of just saying cache and caching makes it a lot more beginner friendly, but worry that we might be setting them up for confusion between the HTTP cache (where all web content is cached) and the Service Worker cache / Cache Storage (where only items from Workbox/Service Worker scripts get cached).

**Comment by addyosmani** [comment]
@kaycebasques reviewed. Very minor nits, but in general I love it ❤️

File: src/index.md
Comment (kaycebasques) [reply src/index.md]: Good call. That definitely sounds more natural

File: src/index.md
Comment (kaycebasques) [reply src/index.md]: Sure, I can do that

For my own understanding: Isn't there a use case for downloading resources before a user needs them? So even though the very first visit on the homepage (for example) will be constrained by a network, you could speed up the first time experience on whatever page the user visits next... I guess that's a repeat-visit experience, even though it's a first-time experience of that page for the user (this is all just semantics)

**Comment by kaycebasques** [comment]
@jeffposnick @gauntface plz leave comments in the markdown if you've got any feedback

**Comment by kaycebasques** [comment]
and @philipwalton

**Comment by jeffposnick** [comment]
Re: pretty URLs, let's keep them turned off and always include the `.html` extension in our links, as we ran into issues with our service worker caching when we had pretty URLs enabled in our Firebase config (see https://github.com/GoogleChrome/workbox-microsite/pull/67)

File: src/overview.md
Comment (jeffposnick) [inline src/overview.md]: Honestly, there isn't a performance benefit of using Workbox vs. using proper HTTP caching headers when it comes to loading subresources used a page. What Workbox lets you do that you couldn't do otherwise is fulfill the navigation request for the initial HTML document using the cache. Could this be rephrased to something like:

> Even when there is a network connection, respond with your site's HTML immediately, instead of having to wait for a response from your server.

File: src/overview.md
Comment (jeffposnick) [inline src/overview.md]: I'd remove this line about the difference. Again, if we're pointing out things that you could do with Workbox that you couldn't do traditionally, the distinction is that you can now respond to navigation requests without having to go to the network.

File: src/overview.md
Comment (jeffposnick) [inline src/overview.md]: Link to https://developers.google.com/web/fundamentals/getting-started/primers/service-workers here?

File: src/overview.md
Comment (jeffposnick) [inline src/overview.md:35]: Being able to precache things is definitely a benefit. But as mentioned in the earlier comments, the main benefit from a performance point of view is being able to use a cache-first approach to serving HTML documents, which cuts out the network latency involved in waiting for a server to respond to a navigation request.

File: src/overview.md
Comment (jeffposnick) [inline src/overview.md]: Could we make the distinction that `workbox-sw` is the runtime library, and we expect that most developers will end up using it, either directly or indirectly via the generated code.

The distinction is that only **one** of `workbox-build`, `workbox-webpack-plugin` or `workbox-cli` would need to be used—the developer would choose the one most appropriate to their build environment. The way it's written, it makes it sound like all the 4 modules do the same thing, and that a developer might end up using any/all of them.

File: src/overview.md
Comment (jeffposnick) [inline src/overview.md:67]: Ideally, "build-time configurations" would link to something meaningful. I think I need to get https://github.com/GoogleChrome/workbox/issues/695 fixed and published in order to have a good explanation, though.

File: src/overview.md
Comment (jeffposnick) [inline src/overview.md]: There's a distinction here—Workbox will manage updated to **precached** resources. Resources that are cached at runtime (like API responses, or optional images) aren't managed in the same way. The update story for those resources depends on which runtime caching strategy is defined for their corresponding routes.

I think that's worth calling out, because there's a central distinction between the build-time precaching functionality vs. runtime caching functionality.

File: src/overview.md
Comment (jeffposnick) [inline src/overview.md]: `workbox-build` is a generic JS module, which could be used from Gulp, but could also be used from other things (Grunt, Broccoli, your-random-JS-build-system).

You can rephrase as:

> `workbox-build` for any node-based build tool, like Gulp or Grunt,

File: src/overview.md
Comment (jeffposnick) [inline src/overview.md]: Can you link to https://workboxjs.org/reference-docs/latest/module-workbox-build.html#.injectManifest to explain how someone would go about "injecting" the manifest?

File: src/overview.md
Comment (jeffposnick) [inline src/overview.md]: Can you link to https://workboxjs.org/reference-docs/latest/module-workbox-build.html#.generateSW to explain what the options are for the "generate" flow?

**Review by jeffposnick — CHANGES_REQUESTED** [review]

File: src/overview.md
Comment (gauntface) [reply src/overview.md]: I'd say this is fine as is. HTTP cache can be wiped out whenever and in that case this will import load perf on repeat visits.

I'd keep this intro as simple to follow and highlevel as possible.

File: src/overview.md
Comment (gauntface) [reply src/overview.md]: +2 removing this sentence.

File: src/overview.md
Comment (gauntface) [reply src/overview.md:35]: I'd still stick with this as is. While there are technicalities to it, I think it gets a core point across and thats what we need to achieve from this doc.

**Review by gauntface — APPROVED** [review]
Generally LGTM with most of Jeffy's comments.

File: src/overview.md
Comment (kaycebasques) [reply src/overview.md]: Done

File: src/overview.md
Comment (kaycebasques) [reply src/overview.md]: Done

File: src/overview.md
Comment (kaycebasques) [reply src/overview.md]: I'll update the next section to include more of this nuance, but I'm going to leave this first section as is. The goal is to concisely explain the use cases / benefits of Workbox.

File: src/overview.md
Comment (kaycebasques) [reply src/overview.md]: Actually, it was unnecessary to list them in this paragraph, so I removed them, but I'll update a later section to address this

File: src/overview.md
Comment (kaycebasques) [reply src/overview.md:67]: OK, we can revisit later. I think the paragraph gets the point across that a declarative solution is often less work than an imperative one

File: src/overview.md
Comment (kaycebasques) [reply src/overview.md]: I updated it to mention "precached resources" instead of "cached" ones, but I omitted the caveat about runtime caching. I'm just trying to convey a high-level understanding of Workbox's benefits. I think this distinction is getting a little too much into the weeds.

File: src/overview.md
Comment (kaycebasques) [reply src/overview.md]: Done

File: src/overview.md
Comment (kaycebasques) [reply src/overview.md]: Done

File: src/overview.md
Comment (kaycebasques) [reply src/overview.md]: Done (it now links below the code)

**Comment by kaycebasques** [comment]
@jeffposnick let me know if the updates are good enough to meet your feedback points

**Review by jeffposnick — APPROVED** [review]

**Comment by jeffposnick** [comment]
I'm 👍 

Thanks again for writing this up!

**Comment by kaycebasques** [comment]
Somebody want to give me write access so I can merge PRs (after reviews)?

**Comment by jeffposnick** [comment]
Looks like you've already got it, @kaycebasques (Someone else might have just added you?)

**Comment by kaycebasques** [comment]
@jeffposnick nope 

screen shot 2017-07-26 at 2 07 18 pm

**Comment by addyosmani** [comment]
@kaycebasques Just gave you admin privs. Could you refresh and check if that fixed it?

**Comment by kaycebasques** [comment]
We gooooood

**Comment by addyosmani** [comment]
giphy

**Pull request merged**
